test/consumer/index-spec.js

```diff
@@ -623,4 +619,59 @@ describe('Consumer tests', function () {
       assert.deepEqual(receivedMessages, ['first', 'second', 'third', 'second', 'third']);
     });
   });
+  describe('Consumer State Change Handling Tests', function () {
+    it('Should run custom state change function provided.', async function () {
+      const stateChanged = await new Promise(async (resolve, reject) => {
+        const stateChangeConsumer = new Consumer({
+          discoveryServers,
+          jwt,
+          topic: 'persistent://public/default/test',
+          subscription: 'subscription',
+          subType: Consumer.SUB_TYPES.FAILOVER,
+          consumerName: 'Consy6',
+          readCompacted: false,
+          receiveQueueSize,
+          logLevel: LEVELS.INFO,
+          stateChangeHandler: ({ previousState, newState }) => {
+            resolve(true);
+          },
+        });
+        consumers.push(stateChangeConsumer);
+        // triggers state change
+        await stateChangeConsumer.subscribe();
+      });
+      assert.ok(stateChanged);
+    });
+    it('Should continue reading even if custom consumer state change function throws errors.', async function () {
+      let expectedNumOfMessages = 20;
+      let actualNumOfMessages = 0;
+      const messages = Array(expectedNumOfMessages).fill('message');
+      await utils.produceMessages({ messages });
+      await new Promise(async (resolve, reject) => {
+        const stateChangeErrorConsumer = new Consumer({
+          discoveryServers,
+          jwt,
+          topic: 'persistent://public/default/test',
+          subscription: 'subscription',
+          subType: Consumer.SUB_TYPES.FAILOVER,
+          consumerName: 'Consy7',
+          readCompacted: false,
+          receiveQueueSize,
+          logLevel: LEVELS.TRACE,
+          stateChangeHandler: ({ previousState, newState }) => {
+            throw new Error('Unexpected fake error!');
+          },
+        });
+        consumers.push(stateChangeErrorConsumer);
+        // triggers state change
```

Comment:
You should trigger the state change while reading messages to make sure there aren't issues such with commits and resending messages to the consumer